### PR TITLE
Rework river generation

### PR DESCRIPTION
**Pull request by Diabolus-Engi**
#### Summary

```SUMMARY: Bugfixes "Reworks river generation on overmap."```

#### Purpose of change

Reworks river generation so that they cross overmap boundaries better and flow more realistically. 

Fixes: #36630
Fixes: #38036

#### Describe the solution

1. Adds river nodes for the overmap to track entry/exit points of rivers so that they can be continued with more accuracy to adjacent overmaps. Also includes the size of the river. This provides some method continuity for rivers and can later be used to maintain other properties for the river such as fish populations, contaminates perhaps.

2. Reworks plotting of the river and generation by using a quadratic bezier curve, coupled with part of the original weighted random walk method that helps to create varied river sizes while the bezier curve allows bends in the river. The control points probably still need some work, but hoping to catch the eye of someone who knows more math than myself for that.

3. Changes the polishing of rivers to apply terrain mask based on adjacent tiles. I found this implementation to be a little cleaned than the original.

4. Rivers have a chance to generate branches of a smaller scale that will merge into the river further down and also generate branches to a random nearby point.

5. Rivers will 'pool up' creating small lakes if they haven't terminated on the overmap boundary.

**TODO:** 
- Remember river scales between overmaps to allow different river sizes to cross.
- Add streams, smaller versions of rivers. Requires mapgen work in a future PR.

#### Describe alternatives you've considered

Well, I originally set out just to fix overmap river borders. After many days of wrestling with the current implementation and testing various alternative methods, I found it was unfeasible to continue without rewriting a lot of the river generation.

Once elevations are added, then we can advance further into erosion models etc that would provide realistic rivers. While I played with the idea of using perlin noise, I opted against it in favor of waiting for elevations.

#### Testing

Generate a new map, teleport to adjacent overmaps and follow the rivers between overmaps.

#### Additional context
(Slightly outdated pre-random branches)
o_10

(Updated)
o_10

### CODE REVIEW

File: src/overmap.cpp [hunk 1]
```diff
@@ -2479,157 +2479,151 @@ void overmap::place_lakes()
     }
 }
 
+bool overmap::is_river_node( const point &p ) const
+{
+    overmap_river_node *n = &get_river_node_at( p );
+    return ( n->p1 == p && n->p2 == p ) ? false : true;
```
Comment (Night-Pryanik) [inline src/overmap.cpp]:
```suggestion
    return !( n->p1 == p && n->p2 == p );
```

File: src/overmap.cpp [hunk 2]
```diff
@@ -2479,157 +2479,151 @@ void overmap::place_lakes()
...
-                is_river( north->ter( p_neighbour + point_west ) ) ) {
-                if( one_in( river_chance ) && ( river_start.empty() ||
-                                                river_start[river_start.size() - 1].x < ( i - 6 ) * river_scale ) ) {
-                    river_start.push_back( p_mine.xy() );
-                }
+
+            if( north->is_river_node( p_neighbour.xy() ) ) {
+                river_start.push_back( p_mine.xy() );
+            } else {
+
```
Comment (Night-Pryanik) [inline src/overmap.cpp]: Nothing on `else` check?
```suggestion
```

File: src/overmap.cpp [hunk 3]
```diff
@@ -2479,157 +2479,151 @@ void overmap::place_lakes()
...
-                is_river( west->ter( p_neighbour + point_south ) ) ) {
-                if( one_in( river_chance ) && ( river_start.size() == rivers_from_north ||
-                                                river_start[river_start.size() - 1].y < ( i - 6 ) * river_scale ) ) {
-                    river_start.push_back( p_mine.xy() );
-                }
+
+            if( west->is_river_node( p_neighbour.xy() ) ) {
+                river_start.push_back( p_mine.xy() );
+            } else {
+
```
Comment (Night-Pryanik) [inline src/overmap.cpp]: The same here.
```suggestion
```

File: src/overmap.cpp [hunk 4]
```diff
@@ -2479,157 +2479,147 @@ void overmap::place_lakes()
...
-        if( river_start.size() != river_end.size() ) {
-            river_start.push_back( point( rng( OMAPX / 4, ( OMAPX * 3 ) / 4 ),
-                                          rng( OMAPY / 4, ( OMAPY * 3 ) / 4 ) ) );
-        }
-        for( size_t i = 0; i < river_start.size(); i++ ) {
-            place_river( river_start[i], river_end[i] );
-        }
+    }
+
+    for( int i = 0; i < river_start.size(); i++ ) {
```
Comment (anothersimulacrum) [inline src/overmap.cpp]:
```suggestion
    for( size_t i = 0; i < river_start.size(); i++ ) {
```

File: src/overmap.cpp [hunk 5]
```diff
@@ -2747,90 +2737,177 @@ void overmap::place_roads( const overmap *north, const overmap *east, const over
...
+    size_t size = 0;
+
+    if( river_scale <= 0 ) {
+        return;
+    }
+
+    const auto cubic_bezier = [&]( point & pa, point & pb, point & pc, point & pd, const int n_segs ) {
+        std::vector<point> pts;
+        for( int i = 0; i <= n_segs; ++i ) {
+            double t = ( double )i / ( double )n_segs;
```
Comment (anothersimulacrum) [inline src/overmap.cpp]:
```suggestion
            double t = static_cast<double>( i ) / static_cast<double>( n_segs );
```

File: src/overmap.cpp [hunk 6]
```diff
@@ -2747,90 +2737,177 @@ void overmap::place_roads( const overmap *north, const overmap *east, const over
...
+
+    sub_ends = cubic_bezier( pa, cpmap1, cpmap2, pb, n_segs );
+    sub_ends.insert( sub_ends.begin(), pa );
+    sub_ends.push_back( pb );
+
+    std::vector<point> tmp;
+    std::vector<point> center_points; //Contains center line of river
+    point last;
+    int orig_scale = river_scale;
+    for( int i = 0; i < sub_ends.size(); i++ ) {
```
Comment (anothersimulacrum) [inline src/overmap.cpp]:
```suggestion
    for( size_t i = 0; i < sub_ends.size(); i++ ) {
```

**Comment by akirashirosawa** [comment]
Is this fixes #39382?

**Comment by ralreegorganon** [comment]
> Does this fix #39382?
No, it's entirely unrelated. This PR changes how the rivers are placed, but that issue is about the local mapgen for lake shores in that particular overmap layout.

**Pull request closed without merging**